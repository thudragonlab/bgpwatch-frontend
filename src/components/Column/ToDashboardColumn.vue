<template>
  <div>
    <div class="clickable" @click="toDashboard">{{ rowFormat(row) }}</div>
  </div>
</template>

<script>
export default {
  name: 'ToDashboardColumn',
  created() {},
  data() {
    return {}
  },
  props: {
    row: {
      type: Object,
    },
    tableParam: {
      type: Object,
    },
    rowFormat: {
      type: Function,
    },
  },
  methods: {
    toDashboard() {
      this.$router.push({ path: '/dashBoard', query: { asn: this.rowFormat(this.row) } }).then(() => {
        location.reload()
      })
    },
  },
}
</script>

<style lang="scss" scoped>
.clickable {
  cursor: pointer;
  color: #2440b3;
  text-decoration: underline;
}
.clickable:hover {
  color: #315efb;
}
</style>
