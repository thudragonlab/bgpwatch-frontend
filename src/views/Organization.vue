<template>
  <div class="org-container">
    <el-row class="search-bar">
      <el-col :col="24">
        <el-input
          id="searchInput"
          :placeholder="`Please input ${placeholderMap[selectedType]}`"
          @input="inputContent"
          v-model="inputText"
        >
          <el-select v-model="selectedType" id="searchSelect" slot="prepend" style="width: 150px" @change="changeModel">
            <el-option v-for="item in options" :key="item.value" :label="item.label" :value="item.value"> </el-option>
          </el-select>
          <el-button slot="append" icon="el-icon-search" @click="doSearch"></el-button>
        </el-input>
      </el-col>
    </el-row>

    <div class="as-table-box">
      <div class="as-table">
        <pagination-table hiddenIndex :originData="tabldata" :showRule="showRule" loadDataFadeIn></pagination-table>
      </div>
    </div>
    <div class="update-time">LastUpdate:{{ lastUpdateTimestamp }}</div>
  </div>
</template>

<script>
import { getAsInfo } from '@/api/routeMonitor'
import ToDashboardColumn from '@/components/Column/ToDashboardColumn.vue'
export default {
  name: 'Organization',
  created() {},
  data() {
    return {
      tabldata: [],
      inputText: 'CERNET',
      lastUpdateTimestamp: '',
      options: [
        {
          value: 'searchASN',
          label: 'ASN',
        },
        {
          value: 'searchAsName',
          label: 'AS Name',
        },
        {
          value: 'searchOrg',
          label: 'Org Name',
        },
      ],
      showRule: [
        {
          prop: 'asn',
          label: 'asn',
          subComponent: ToDashboardColumn,
          rowFormat: row => {
            return row.asn
          },
        },

        {
          prop: 'as_name',
          label: 'AS Name',
        },
        {
          prop: 'org_name',
          label: 'Organization',
        },
      ],
      selectedType: 'searchAsName',
      searchASN: '',
      searchAsName: 'CERNET',
      searchOrg: '',
      propList: ['searchASN', 'searchAsName', 'searchOrg'],
      placeholderMap: {
        searchASN: 'asn',
        searchAsName: 'as name',
        searchOrg: 'as organization',
      },
    }
  },
  async mounted() {
    this.doSearch()
  },
  methods: {
    async doSearch() {
      const result = await getAsInfo(this.searchASN, this.searchAsName, this.searchOrg)
      this.lastUpdateTimestamp = new Date(parseInt(localStorage.getItem('lastUpdateTimestamp') * 1000))
        .toJSON()
        .split('T')[0]
      this.tabldata = result
    },
    changeModel(v) {
      this.inputText = ''
      this.propList.forEach(prop => {
        if (v !== prop) {
          this[prop] = ''
        }
      })
    },
    inputContent() {
      this[this.selectedType] = this.inputText
    },
  },
}
</script>

<style lang="scss" scoped>
.org-container {
  padding: 22px;
  background-color: #f0f0f0;
  height: 100%;
}

.as-table-box {
  border-radius: 20px;
  padding: 10px;
  background-color: #fff;
}

.as-table {
  // border-radius: 20px;
}

.search-bar {
  background-color: #fff;
  border-radius: 20px;
  padding: 10px;
}

.update-time {
  background-color: #fff;
  text-align: right;
  font-weight: bold;
  margin-top: 10px;
  padding-right: 10px;
  border-radius: var(--border-radius-deg);
}
</style>
