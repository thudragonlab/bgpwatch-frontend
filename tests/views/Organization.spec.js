import {
  mount,
  shallowMount
} from "@vue/test-utils";
import ElementUI from '@/lib/element.js'
import ToDashboardColumn from '@/components/Column/ToDashboardColumn.vue'

import Organization from "@/views/Organization";

import MyC from 'dragonlab-component/lib/dragonLab-Component.umd.js'
import Vue from "vue";
import * as testData from '../mockData/Organization'

import {
  getAsInfo
} from '@/api/routeMonitor'

jest.mock('@/api/routeMonitor', () => {
  // const _inner = [...testData]
  const mock = require('../mockData/Organization')
  return {
    getAsInfo: jest.fn(() => Promise.resolve(mock.default))
  }
})

Vue.use(MyC)
Vue.use(ElementUI);
Vue.component('subComponent0', ToDashboardColumn)


describe('Organization-HTML', () => {

  it('test all event', async () => {
    const wrapper = mount(Organization, {
      stubs: {
        transition: false
      },
    })
    expect(wrapper.html()).toMatchSnapshot();
    const inputContent = jest.spyOn(wrapper.vm,'inputContent')
    const changeModel = jest.spyOn(wrapper.vm,'changeModel')
    const doSearch = jest.spyOn(wrapper.vm,'doSearch')
    // wrapper.vm.inputContent = inputContent
    const input = wrapper.find('#searchInput')
    const inputBox = wrapper.find('.el-input')
    
    await input.setValue('4262')
    inputBox.vm.$emit('input','4262')
    
    // input.vm.$emit('input')
    expect(inputContent).toHaveBeenCalled()
    expect(wrapper.vm.inputText).toBe('4262')
    expect(wrapper.vm[wrapper.vm.selectedType]).toBe('4262')

    const select = wrapper.find('#searchSelect')
    const selectBox = wrapper.find('.el-select')

    await select.setValue('searchASN')
    selectBox.vm.$emit('change','searchASN')
    await Vue.nextTick()
    expect(wrapper.vm.inputText).toBe('')
    wrapper.vm.propList.forEach(prop => {
      expect(wrapper.vm[prop]).toBe('')
    })

    expect(changeModel).toHaveBeenCalledWith('searchASN')

    const button = wrapper.find('.el-button')
    await button.trigger('click')

    expect(doSearch).toHaveBeenCalled()
    
  })
})

// describe('Organization-HTML.vue', () => {

//   it('searches ASNs by ASN, AS name, or organization', async () => {
//     // Arrange
//     const wrapper = mount(Organization, { 
//       mocks: {
//         $nextTick: fn => fn(),
//         $message: () => {},
//       },
//     });
//     expect(wrapper.html()).toMatchSnapshot();
//     // Act
//     wrapper.find('[placeholder="Please input asn"]').setValue('4262')
//     wrapper.find('[placeholder="Please input asn"]').trigger('input')
//     await wrapper.vm.doSearch();
//     let rows = wrapper.findAll('.el-table__row');
//     expect(rows.length).toBe(3);
//     expect(rows.at(0).text()).toContain('4262');

//     // wrapper.setData({ searchAsName: 'CERNET' });
//     wrapper.find('[placeholder="Please input as name"]').setValue('CERNET')
//     wrapper.find('[placeholder="Please input as name"]').trigger('input')
//     await wrapper.vm.doSearch();
//     rows = wrapper.findAll('.el-table__row');
//     expect(rows.length).toBe(3);
//     expect(rows.at(0).text()).toContain('4262');
//     expect(rows.at(1).text()).toContain('4263');
//     expect(rows.at(2).text()).toContain('4264');

//     // wrapper.setData({ searchOrg: 'California Education' });
//     wrapper.find('[placeholder="Please input as organization"]').setValue('California Education')
//     wrapper.find('[placeholder="Please input as organization"]').trigger('input')
//     await wrapper.vm.doSearch();
//     rows = wrapper.findAll('.el-table__row');
//     expect(rows.length).toBe(3);
//     expect(rows.at(0).text()).toContain('4262');
//     expect(rows.at(1).text()).toContain('4263');
//     expect(rows.at(2).text()).toContain('4264');

//     // Assert
//     expect(getAsInfo).toHaveBeenCalledTimes(4);
//     expect(getAsInfo).toHaveBeenCalledWith('', 'CERNET', '');
//     expect(getAsInfo).toHaveBeenCalledWith('4262', '', '');
//     expect(getAsInfo).toHaveBeenCalledWith('', 'CERNET', '');
//     expect(getAsInfo).toHaveBeenCalledWith('', '', 'California Education');
//   });
// });

describe('Organization.vue', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(Organization)
  })
  afterEach(() => {
    wrapper.destroy()
  })
  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true)
  })
  it('calls getAsInfo on mount', async () => {
    await wrapper.vm.$nextTick()
    expect(getAsInfo).toHaveBeenCalledWith('', 'CERNET', '')
  })
  it('calls getAsInfo on doSearch', async () => {
    wrapper.vm.searchASN = '123'
    wrapper.vm.searchOrg = 'ABC'
    await wrapper.vm.doSearch()
    expect(getAsInfo).toHaveBeenCalledWith('123', 'CERNET', 'ABC')
    expect(wrapper.vm.tabldata).toEqual([{
        "as_name": "CERNET-ASN-BLOCK (US)",
        "asn": "4262",
        "org_name": "California Education and Research Federation Network"
      },
      {
        "as_name": "CERNET-ASN-BLOCK (US)",
        "asn": "4263",
        "org_name": "California Education and Research Federation Network"
      },
      {
        "as_name": "CERNET-ASN-BLOCK (US)",
        "asn": "4264",
        "org_name": "California Education and Research Federation Network"
      }
    ])
  })
  it('inputContent', () => {
    wrapper.vm.selectedType = 'searchASN'
    wrapper.vm.inputText = '123'
    wrapper.vm.inputContent()
    expect(wrapper.vm.searchASN).toBe('123')
  })

  it('changeModel', () => {
    wrapper.vm.selectedType = 'searchASN'
    wrapper.vm.searchASN = '123'
    wrapper.vm.searchAsName = '123'
    wrapper.vm.searchOrg = '123'
    wrapper.vm.changeModel(wrapper.vm.selectedType)
    expect(wrapper.vm.searchASN).toBe('123')
    expect(wrapper.vm.searchOrg).toBe('')
    expect(wrapper.vm.searchAsName).toBe('')
    expect(wrapper.vm.inputText).toBe('')
  })



})