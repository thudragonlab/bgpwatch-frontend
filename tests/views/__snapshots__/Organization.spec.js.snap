// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organization-HTML test all event 1`] = `
<div class="org-container">
  <div class="search-bar el-row">
    <div class="el-col el-col-24" col="24">
      <div class="el-input el-input-group el-input-group--append el-input-group--prepend">
        <div class="el-input-group__prepend">
          <div class="el-select" style="width: 150px;">
            <!---->
            <div class="el-input el-input--suffix">
              <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="Select" id="searchSelect" class="el-input__inner">
              <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----><!----><!----><!----></span>
              <!----></span>
              <!---->
              <!---->
            </div>
            <div class="el-select-dropdown el-popper" style="display: none;">
              <div class="el-scrollbar" style="display: none;">
                <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                  <ul class="el-scrollbar__view el-select-dropdown__list">
                    <!---->
                    <li class="el-select-dropdown__item"><span>ASN</span></li>
                    <li class="el-select-dropdown__item selected"><span>AS Name</span></li>
                    <li class="el-select-dropdown__item"><span>Org Name</span></li>
                  </ul>
                </div>
                <div class="el-scrollbar__bar is-horizontal">
                  <div class="el-scrollbar__thumb" style="width: 0px; transform: translateX(0%); webkit-transform: translateX(0%);"></div>
                </div>
                <div class="el-scrollbar__bar is-vertical">
                  <div class="el-scrollbar__thumb" style="height: 0px; transform: translateY(0%); webkit-transform: translateY(0%);"></div>
                </div>
              </div>
              <p class="el-select-dropdown__empty">
                No data
              </p>
            </div>
          </div>
        </div><input type="text" autocomplete="off" id="searchInput" placeholder="Please input as name" class="el-input__inner">
        <!---->
        <!---->
        <div class="el-input-group__append"><button type="button" class="el-button el-button--default">
            <!----><i class="el-icon-search"></i>
            <!----></button></div>
        <!---->
      </div>
    </div>
  </div>
  <div class="as-table-box">
    <div class="as-table">
      <div>
        <!---->
        <div class="transition el-row">
          <div class="el-table el-table--fit el-table--striped el-table--border el-table--enable-row-hover" style="width: 100%; --text: 10;">
            <div class="hidden-columns">
              <!---->
              <div subcomponent="[object Object]" rowformat="row => {
          return row.asn;
        }"></div>
              <div></div>
              <div></div>
            </div>
            <div class="el-table__header-wrapper">
              <table cellspacing="0" cellpadding="0" border="0" class="el-table__header">
                <colgroup></colgroup>
                <thead class="">
                  <tr class=""></tr>
                </thead>
              </table>
            </div>
            <div class="el-table__body-wrapper is-scrolling-none">
              <table cellspacing="0" cellpadding="0" border="0" class="el-table__body">
                <colgroup></colgroup>
                <tbody>
                  <!---->
                </tbody>
              </table>
              <div class="el-table__empty-block" style="height: 100%;"><span class="el-table__empty-text">No Data</span></div>
              <!---->
            </div>
            <!---->
            <!---->
            <!---->
            <!---->
            <div class="el-table__column-resize-proxy" style="display: none;"></div>
          </div>
          <div class="pagination">
            <div class="el-pagination"><button type="button" disabled="disabled" class="btn-prev"><i class="el-icon el-icon-arrow-left"></i></button>
              <ul class="el-pager">
                <li class="number active">1</li>
                <!---->
                <!---->
                <!---->
              </ul><button type="button" disabled="disabled" class="btn-next"><i class="el-icon el-icon-arrow-right"></i></button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="update-time">LastUpdate:</div>
</div>
`;
